### ISSUE TITLE
"duplicate token" error in web admin UI in certain browsers

### ISSUE THREAD

**Issue by tomxtobin** [rethinkdb/rethinkdb#4174]
When loading the web admin UI at `localhost:8080` in a couple of different WebKit-based browsers (qutebrowser and dwb), I get the following error for a few seconds instead of the status, performance, and log widgets:

```
Error

The request to retrieve data failed. You can try to refresh.

Error:

ERROR: duplicate token 1 in:
r.do(r.db("rethinkdb").table("table_status").coerceTo("array"), r.db("rethinkdb").table("table_config", {identifierFormat: "uuid"}).coerceTo("array"), r.db("rethinkdb").table("server_config").coerceTo("array"), r.db("rethinkdb").table("server_status").coerceTo("array"), r.db("rethinkdb").table("jobs").coerceTo("array"), r.db("rethinkdb").table("current_issues").coerceTo("array"), function(var_2, var_3, var_4, var_5, var_6, var_7) { return {num_primaries: var_3("shards").map(function(var_8) { return var_8.count(); }).sum(), num_connected_primaries: var_2.map(function(var_9) { return var_9("shards")("primary_replica").count(function(var_10) { return var_10.ne(null); }); }).sum(), num_replicas: var_3("shards").map(function(var_11) { return var_11.map(function(var_12) { return var_12("replicas").count(); }).sum(); }).sum(), num_connected_replicas: var_2("shards").map(function(var_13) { return var_13("replicas").map(function(var_14) { return var_14("state").count(function(var_15) { return r(["ready", "looking_for_primary_replica"]).contains(var_15); }); }).sum(); }).sum(), tables_with_primaries_not_ready: r.db("rethinkdb").table("server_config").map(function(var_16) { return [var_16("id"), var_16("name")]; }).coerceTo("ARRAY").coerceTo("OBJECT").do(function(var_17) { return var_2.map(var_3, function(var_18, var_19) { return {id: var_18("id"), name: var_18("name"), db: var_18("db"), shards: var_18("shards").map(r.range(), var_19("shards"), function(var_20, var_21, var_22) { return {position: var_21.add(1), num_shards: var_18("shards").count(), primary_id: var_22("primary_replica"), primary_name: var_17(var_22("primary_replica")).default("-"), primary_state: var_20("replicas").filter({server: var_17(var_22("primary_replica"))}, {default: r.error()})("state")(0).default("missing")}; }).filter(function(var_23) { return r(["ready", "looking_for_primary_replica"]).contains(var_23("primary_state")).not(); }).coerceTo("array")}; }).filter(function(var_24) { return var_24("shards").isEmpty().not(); }).coerceTo("array"); }), tables_with_replicas_not_ready: var_2.map(var_3, function(var_25, var_26) { return {id: var_25("id"), name: var_25("name"), db: var_25("db"), shards: var_25("shards").map(r.range(), var_26("shards"), function(var_27, var_28, var_29) { return {position: var_28.add(1), num_shards: var_25("shards").count(), replicas: var_27("replicas").filter(function(var_30) { return r(["ready", "looking_for_primary_replica", "offloading_data"]).contains(var_30("state")).not(); }).map(var_29("replicas"), function(var_31, var_32) { return {replica_id: var_32, replica_name: var_31("server")}; }).coerceTo("array")}; }).coerceTo("array")}; }).filter(function(var_33) { return var_33("shards")(0)("replicas").isEmpty().not(); }).coerceTo("array"), num_tables: var_3.count(), num_servers: var_5.count(), num_connected_servers: var_5.count(function(var_34) { return var_34("status").eq("connected"); }), disconnected_servers: var_5.filter(function(var_35) { return var_35("status").ne("connected"); }).map(function(var_36) { return {time_disconnected: var_36("connection")("time_disconnected"), name: var_36("name")}; }).coerceTo("array"), num_disconnected_tables: var_2.count(function(var_37) { return var_37("shards").map(function(var_38) { return var_38("primary_replica").eq(null); }).contains(true); }), num_tables_w_missing_replicas: var_2.count(function(var_39) { return var_39("status")("all_replicas_ready").not(); }), num_sindex_issues: var_7.count(function(var_40) { return var_40("type").eq("outdated_index"); }), num_sindexes_constructing: var_6.count(function(var_41) { return var_41("type").eq("index_construction"); })}; })
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

![rethinkdb-error](https://cloud.githubusercontent.com/assets/28852/7437274/2b906d68-f024-11e4-9061-e84ffc376d18.jpg)

The UI then seems to toggle between displaying the widgets, and the error (or some variant of it, with a different number N for `duplicate token N`).

Sometimes it only displays the error inside of the log widget:

![rethinkdb-error](https://cloud.githubusercontent.com/assets/28852/7437311/6def2ab4-f024-11e4-96d0-b13af14a9da6.jpg)

This doesn't happen in Firefox or Chromium, for what it's worth. I'm running RethinkDB 2.0.1 under Arch Linux.

**Comment by deontologician** [comment]
Hi @tomxtobin, I'm checking this out now. Thanks for reporting it

**Comment by deontologician** [comment]
I was able to replicate this in qutebrowser and midori.

Looks like this is due to absolutely stubborn caching of XHR responses. No combination of `Cach-Control: no-cache`, `Pragma: no-cache`, or `Expires` headers could convince these browsers not to cache the POST request that creates a new connection. I ended up adding a random query parameter to the url to finally fix it.

As part of this, I found another bug, which is that our code doesn't correctly write tokens > `2^31` to the connection. Not sure if there's any reason to fix that particular bug since 2 billion queries on a single connection is a lot, but it's something we should probably at least have an issue for.

The fix is up in CR 2899 by @AtnNn
